luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,160 @@
+local updateInterval = Game.gameSpeed / 2
+
+local spValidUnitID = Spring.ValidUnitID
+local spValidFeatureID = Spring.ValidFeatureID
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetUnitSeparation = Spring.GetUnitSeparation
+local spGetUnitFeatureSeparation = Spring.GetUnitFeatureSeparation
+local spGetUnitCommands = Spring.GetUnitCommands
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+--- Key is the nano's ID, followed by a set-like with the unitID of target currently inside the nano's range.
+--- Used for CommandArray manipulation
+---@type {[integer] : {integer : {cmdTag : integer, status : integer}}}
+local trackingTable = {}
+-- Limits how many GetUnitCommands are called inside a frame when required.
+-- higher values allow faster tagging synchronization after a new command;
+-- which can cause proportional lag when many units are targeted. (long tables returned)
+local tagUpdateFrameMax = 1
+
+local statuses = {
+	outOfRange = 0,
+	inRange = 1,
+}
+
+local mobileUnits = {}
+local constructionTurretsDefs = {}
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
+		}
+	end
+
+	if unitDef.canMove then
+		mobileUnits[unitDefID] = true
+	end
+end
+
+function gadget:Initialize()
+	gadgetHandler:RegisterAllowCommand(CMD.REPAIR)
+	gadgetHandler:RegisterAllowCommand(CMD.GUARD)
+	gadgetHandler:RegisterAllowCommand(CMD.RECLAIM)
+	gadgetHandler:RegisterAllowCommand(CMD.STOP)
+end
+
+function gadget:GameFrame(frame)
+	local tagUpdates = 0
+	local cmdCache = {}
+
+	for nanoID, targets in pairs(trackingTable) do
+		if nanoID % updateInterval == frame % updateInterval then
+			if next(targets) == nil then
+				trackingTable[nanoID] = nil
+			end
+
+			local nanoDefID = spGetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+
+			for targetID, commandData in pairs(targets) do
+				if spValidUnitID(targetID) then
+					if not commandData.cmdTag then
+						if tagUpdates >= tagUpdateFrameMax then break end
+
+						if not cmdCache[nanoID] then
+							cmdCache[nanoID] = spGetUnitCommands(nanoID, -1)
+							tagUpdates = tagUpdates +1
+						end
+
+						for _, cmd in ipairs(cmdCache[nanoID]) do
+							trackingTable[nanoID][cmd.params[1]].cmdTag = cmd.tag
+						end
+					end
+
+					local distance = spGetUnitSeparation(nanoID, targetID, constructionTurretsDefs[nanoDefID].buildRange3D, false)
+
+					if distance < maxDistance then
+						if commandData.status == statuses.outOfRange then
+							commandData.status = statuses.inRange
+						end
+					else
+						-- Do not blindly remove orders on out-of-range mobile units, as the player may have pre-emptively issued an order before the unit was in range
+						-- Instead, only remove the order once the unit has come into range then left again
```

Comment:
This behaviour where moving units get tracked and removed when out of range could be unconvenient, can't it? like if you have some turrets helping a builder and it goes out of range for a sec, they will just start doing smth else. not sure everyone will like this and can result in unexpected situations for the player.

It could be controlled through a config option later tho, so not a real problem, personally as a player I think I would rather have it off and just have the range control when issuing commands against static units, to avoid having to worry about that.

Reply:
The whole point of this gadget is so that construction turrets don't get stuck on an impossible to fulfil order, causing them to stop doing anything.

Reply:
I really like that for static units, since otherwise the behaviour is buggy, I was even using this widget before when playing bar.

Thing is when working with moving units I don't think it's so clear cut, since I might have a worker doing things assisted by turrets. If he goes away for a sec, like to rebuild smth an enemy just destroyed, and soon comes back he should still be assisted, and it's also not clear I'd want the turrets to assist something else in that case, since that might be my eco builder. Just an example, but I think it can be "too smart" in some situations.

I think in the case of moving builders if the builder comes back the turrets should resume assisting it, maybe move the command back in the queue?. That would be perfect.

Reply:
It sounds like what we really want is the ability to proceed to the next order in the queue without dropping the current one.

Reply:
Maybe but might be a bit far fetched.

Anyways I was just saying, don't think it's a blocker, specially since its a controversial opinion, but still think that behaviour can be inconvenient/unexpected at times and I'm sure it will bite more than one player in the ass.

Reply:
For now, let's leave behaviour as is, then take a second pass once we have some in-game data.

Reply:
So exclude mobile builder from tracking, gotcha.